packages/flutter_tools/lib/src/base/process_manager.dart

```diff
@@ -479,3 +491,294 @@ class _RecordingProcess implements Process {
+///     used in the `stdout` file. If unspecified, then the file was written
+///     as binary data.
+///   - `stderrEncoding` (optional): The name of the encoding scheme that was
+///     used in the `stderr` file. If unspecified, then the file was written
+///     as binary data.
+///   - `exitCode` (required): The exit code of the process, or null if the
+///     process was not responding.
+///   - `daemon` (optional): A boolean indicating that the process is to stay
+///     resident during the entire lifetime of the master Flutter tools process.
+/// - a `stdout` file for each process invocation. The location of this file
+///   can be derived from the `basename` manifest property like so:
+///   `'$basename.stdout'`.
+/// - a `stderr` file for each process invocation. The location of this file
+///   can be derived from the `basename` manifest property like so:
+///   `'$basename.stderr'`.
+class ReplayProcessManager implements ProcessManager {
+  final List<Map<String, dynamic>> _manifest;
+  final Directory _dir;
+
+  ReplayProcessManager._(this._manifest, this._dir);
+
+  /// Creates a new `ReplayProcessManager` capable of replaying a recording at
+  /// the specified location.
+  ///
+  /// If [location] represents a file, it will be treated like a recording
+  /// ZIP file. If it points to a directory, it will be treated like an
+  /// unzipped recording. If [location] points to a non-existent file or
+  /// directory, an [ArgumentError] will be thrown.
+  static Future<ReplayProcessManager> create(String location) async {
+    Directory dir;
+    switch (await FileSystemEntity.type(location)) {
+      case FileSystemEntityType.FILE:
+        dir = await Directory.systemTemp.createTemp('flutter_tools_');
+        os.unzip(new File(location), dir);
+        addShutdownHook(() async {
+          await dir.delete(recursive: true);
+        });
+        break;
+      case FileSystemEntityType.DIRECTORY:
+        dir = new Directory(location);
+        break;
+      case FileSystemEntityType.NOT_FOUND:
+        throw new ArgumentError.value(location, 'location');
+    }
+
+    File manifestFile = new File(path.join(dir.path, _kManifestName));
+    if (!(await manifestFile.exists())) {
+      // We use the existence of the manifest as a proxy for this being a
+      // valid replay directory. Namely, we don't validate the structure of the
+      // JSON within the manifest, and we don't validate the existence of
+      // all stdout and stderr files referenced in the manifest.
+      throw new ArgumentError('$_kManifestName not found.');
+    }
+
+    String content = await manifestFile.readAsString();
+    try {
+      List<Map<String, dynamic>> manifest = new JsonDecoder().convert(content);
+      return new ReplayProcessManager._(manifest, dir);
+    } on FormatException catch (_) {
+      throw new ArgumentError('$_kManifestName is not a valid JSON file.');
+    }
```

Comment:
Maybe `assert(false)` here for future-proofing in case someone starts throwing something other than `FormatException`.